deps/sandbox.c

```diff
@@ -444,6 +444,131 @@ static void mount_the_world(const char * root_dir, const char * dest,
   mount_procfs("");
 }
 
+/*** 3: Networking
+ *
+ * This section sets up networking. To do so, it instructs the kernel
+ * netwokring subsystem to set the appropriate data structures. The ip
+ * address is set using SIOCSIFADDR (set netdevice(7)). The default gateway
+ * is set using the route subsytem (see rtnetlink(7)). The commands we perform
+ * are roughly:
+ *
+ *  // Find out what the name of the network interface is
+ *  ipconfig eth0 10.0.2.3
+ *  ifup eth0
+ *  ip route add default via 10.0.2.2
+ *
+ * But rather than going through the hassle of requiring those external
+ * commands to be present, we just do it through the kernel API, which
+ * while a bit cumbersome is quite useable.
+ */
+void buf_put(char **cur_buf_pos, void *data, size_t size)
+{
+  memcpy(*cur_buf_pos, data, size);
+  *cur_buf_pos += size;
+}
+
+void buf_put_attr_header(char **cur_buf_pos, uint16_t opt, size_t size)
+{
+  struct rtattr attr = {.rta_type = opt,
+                        .rta_len = size + sizeof(struct rtattr)};
+  buf_put(cur_buf_pos, &attr, sizeof(struct rtattr));
+}
+
+void buf_put_attr(char **cur_buf_pos, uint16_t opt, void *data, size_t size)
+{
+  buf_put_attr_header(cur_buf_pos, opt, size);
+  buf_put(cur_buf_pos, data, size);
+  *cur_buf_pos += RTA_ALIGN(size) - size;
+}
+
+#define htonl(x) __bswap_32(x)
```

Comment:
Shouldn't `htonl()` be defined in some header somewhere?

Reply:
Yeah, but it's in one of those headers that redefines a bunch of otherwise used constants and conflicts with things.